src/publisher/message-queues.ts

```diff
@@ -0,0 +1,265 @@
+      },
+      (err, resp) => {
+        const messageIds = (resp && resp.messageIds) || [];
+        callbacks.forEach((callback, i) => callback(err, messageIds[i]));
+        done!(err);
+      }
+    );
+  }
+}
+
+/**
+ * Standard message queue used for publishing messages.
+ *
+ * @private
+ * @extends MessageQueue
+ *
+ * @param {Publisher} publisher The publisher.
+ */
+export class Queue extends MessageQueue {
+  batch: MessageBatch;
+  constructor(publisher: Publisher) {
+    super(publisher);
+    this.batch = new MessageBatch(this.batchOptions);
+  }
+  /**
+   * Adds a message to the queue.
+   *
+   * @param {object} message The message to publish.
+   * @param {PublishCallback} callback The publish callback.
+   */
+  add(message: PubsubMessage, callback: PublishCallback): void {
+    if (!this.batch.canFit(message)) {
+      this.publish();
+    }
+
+    this.batch.add(message, callback);
+
+    if (this.batch.isFull()) {
+      this.publish();
+    }
+
+    if (!this.pending) {
+      const {maxMilliseconds} = this.batchOptions;
+      this.pending = (setTimeout(
+        () => this.publish(),
+        maxMilliseconds
+      ) as unknown) as NodeJS.Timer;
+    }
+  }
+  /**
+   * Cancels any pending publishes and calls _publish immediately.
+   */
+  publish(): void {
+    const {messages, callbacks} = this.batch;
+
+    this.batch = new MessageBatch(this.batchOptions);
+
+    if (this.pending) {
+      clearTimeout(this.pending);
+      delete this.pending;
+    }
+
+    this._publish(messages, callbacks);
+  }
+}
+
+/**
+ * Queue for handling ordered messages. Unless the standard queue, this
+ * ensures that batches are published one at a time and throws an exception in
+ * the event that any batch fails to publish.
+ *
+ * @private
+ * @extends MessageQueue
+ *
+ * @param {Publisher} publisher The publisher.
+ * @param {string} key The key used to order the messages.
+ */
+export class OrderedQueue extends MessageQueue {
+  batches: MessageBatch[];
+  inFlight: boolean;
+  error?: null | ServiceError;
+  key: string;
+  constructor(publisher: Publisher, key: string) {
+    super(publisher);
+    this.batches = [];
+    this.inFlight = false;
+    this.key = key;
+  }
+  /**
+   * Adds a message to a batch, creating a new batch if need be.
+   *
+   * @param {object} message The message to publish.
+   * @param {PublishCallback} callback The publish callback.
+   */
+  add(message: PubsubMessage, callback: PublishCallback): void {
+    let batch = this.batches[0];
+
+    if (!batch || !batch.canFit(message)) {
+      batch = new MessageBatch(this.batchOptions);
+      this.batches.unshift(batch);
+    }
+
+    batch.add(message, callback);
+
+    if (this.inFlight || this.pending) {
+      return;
+    }
+
+    this.beginNextPublish();
+  }
+  /**
+   * Prepares the next batch to be published. If the batch is full we will
+   * publish immediately, otherwise we'll wait until the `maxMilliseconds`
```

Comment:
If there is a publish outstanding, then we should continue to batch until that publish has completed, even if it exceeds the batch time specified in the batchOptions.

Reply:
I believe this does that. While waiting for a publish to finish we continue to build batches based on the `maxMessages` and `maxBytes` options (maybe that is wrong?) and essentially build a queue of batches. Once a publish is complete we check to see if the next batch is either full or if we hit the `maxMilliseconds`.

Thinking about it now, while a pending publish is in place we should continue to batch until the API limit is met, which in turn means we ignore the user limits in this scenario. Does that sound right?